solutions/cses-1163.mdx

```diff
@@ -61,47 +62,152 @@ int main(){
+## Solution 2 - Going Backwards
+	
+We'll start by trying to find the maximum gap once all the traffic lights are
+added. This is the last thing we'll output, so we'll add it to the end of our
+output array. Then, we'll remove traffic lights in the reverse order to how they
+were added, and find the gap each removal creates.
+
+This gap is just the distance between the two street coordinates (either a
+traffic light or the beginning or ending of the street) next to a given traffic
+light stored in our set, so we can use our set to find these values and subtract
+them to get our gap.
+
+But this gap may not be the maximum gap. We'll compare this to the gap we found
+once all the traffic lights are added, and set the max gap to the greater value.
+We'll then set the next lowest element in the output array to this value, which
+will represent the greatest gap prior to adding the traffic light we just
+removed.
+	
+## Implementation
+**Time Complexity:** $\mathcal{O}(n)$
+	
+<LanguageSection>
+<JavaSection>
+
+<Warning>
+We have to use some nonstandard io (input/output) to pass all test cases in the allowed time on CSES, because of its slow grader. 
+</Warning>
+```java
+import java.io.*;
+import java.util.*;
+
+public class TrafficLights {
+	public static void main(String[] args) throws IOException {
+		BufferedReader r = new BufferedReader(new InputStreamReader(System.in));
+		PrintWriter pw = new PrintWriter(System.out);
+		StringTokenizer st = new StringTokenizer(r.readLine());
+
+		int streetLength = Integer.parseInt(st.nextToken());
+		int lightNum = Integer.parseInt(st.nextToken());
+
+		// Using an array to read values since we can't get values from sets in Java
+		int[] opArray = new int[lightNum];
+		NavigableSet<Integer> streetPositions = new TreeSet<>();
+		// Initialize the set with beginning and ending values
+		streetPositions.add(0);
+		streetPositions.add(streetLength);
+		st = new StringTokenizer(r.readLine());
+
+		for (int i = 0; i < lightNum; i++) {
+			int nextTrafficLight = Integer.parseInt(st.nextToken());
+			opArray[i] = nextTrafficLight;
+			streetPositions.add(nextTrafficLight);
+		}
+		int[] gapsArray = new int[lightNum];
+		int prev = 0;
+		int maxGap = 0;
+		// Find the longest passage once all the streetlights are added
+		for (int i : streetPositions) {
+			maxGap = Math.max(i - prev, maxGap);
+			prev = i;
+		}
+		gapsArray[lightNum - 1] = maxGap;
+		/*
+		 * Remove the streetlights in reverse order to how they were added, then find
+		 * the gap created by removing each. Find the biggest current gap, and 
+		 * add it to the next lowest index in answer.
+		 */
+		for (int i = lightNum - 1; i > 0; i--) {
+			streetPositions.remove(opArray[i]);
+			int low = streetPositions.lower(opArray[i]);
+			int high = streetPositions.higher(opArray[i]);
+			maxGap = Math.max(maxGap, high - low);
+			gapsArray[i - 1] = maxGap;
+		}
+		//  Use StringBuilder to print out the array quicker
+		StringBuilder sb = new StringBuilder();
```

Comment:
Not sure if this optimization is needed because printwriter prints out everything at the end with pw.close(), so there isn't really a delay with printing it out one by one in the for loop. Let me know if this is not the case

Reply:
Again, you can probably also do this with kattio since kattio uses printwriter

Reply:
The same as above, Kattio doesn't work fast enough. Is it proper to resolve conversations in this case?